sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/constraints.scala

```diff
@@ -0,0 +1,179 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.expressions
+
+import org.antlr.v4.runtime.ParserRuleContext
+
+import org.apache.spark.sql.catalyst.parser.ParseException
+import org.apache.spark.sql.types.{DataType, StringType}
+
+trait TableConstraint {
+
+  /** Returns the name of the constraint */
+  def name: String
+
+  /** Returns the characteristics of the constraint (e.g., ENFORCED, RELY) */
+  def characteristic: ConstraintCharacteristic
+
+  /** Creates a new constraint with the given name
+   * @param name Constraint name
+   * @return New TableConstraint instance
+   */
+  def withName(name: String): TableConstraint
+
+  /** Creates a new constraint with the given characteristic
+   * @param c Constraint characteristic (ENFORCED, RELY)
+   * @param ctx Parser context for error reporting
+   * @return New TableConstraint instance
+   */
+  def withCharacteristic(c: ConstraintCharacteristic, ctx: ParserRuleContext): TableConstraint
+
+  /** Generates a constraint name if one is not provided
+   * @param tableName Name of the table containing this constraint
+   * @return TableConstraint with a generated name if original name was null/empty
+   */
+  def generateConstraintNameIfNeeded(tableName: String): TableConstraint = {
+    if (name == null || name.isEmpty) {
+      this.withName(name = generateConstraintName(tableName))
+    } else {
+      this
+    }
+  }
+
+  // Generate a constraint name based on the table name if the name is not specified
+  protected def generateConstraintName(tableName: String): String
+}
+
+case class ConstraintCharacteristic(enforced: Option[Boolean], rely: Option[Boolean])
+
+object ConstraintCharacteristic {
+  val empty: ConstraintCharacteristic = ConstraintCharacteristic(None, None)
+}
+
+case class CheckConstraint(
+    child: Expression,
+    condition: String,
+    override val name: String = null,
+    override val characteristic: ConstraintCharacteristic = ConstraintCharacteristic.empty)
+  extends UnaryExpression
+  with Unevaluable
+  with TableConstraint {
+
+  override protected def withNewChildInternal(newChild: Expression): Expression =
+    copy(child = newChild)
+
+  override protected def generateConstraintName(tableName: String): String = {
+    val base = condition.filter(_.isLetterOrDigit).take(20)
+    val rand = scala.util.Random.alphanumeric.take(6).mkString
+    s"${tableName}_chk_${base}_$rand"
+  }
+
+  override def sql: String = s"CONSTRAINT $name CHECK ($condition)"
+
+  override def dataType: DataType = StringType
+
+  override def withName(name: String): TableConstraint = copy(name = name)
+
+  override def withCharacteristic(
+      c: ConstraintCharacteristic,
+      ctx: ParserRuleContext): TableConstraint =
+    copy(characteristic = c)
+}
+
+case class PrimaryKeyConstraint(
+    columns: Seq[String],
```

Comment:
As there is `MULTIPLE_PRIMARY_KEYS`, I think there is only one column in `PrimaryKeyConstraint`?

Reply:
primary key can be a combination of columns, e.g `primary key (a, b)`